### PR TITLE
Use earliest latest compilers

### LINKED ISSUE #7891: Do a build with earliest and latest compilers
The goal of this task is to replace `component_test_gcc_opt` in `tests/scripts/all.sh` by two components that use the compilers `gcc-latest`, and `gcc-earliest`. Also do the same for `clang` (`clang-earliest` and `clang-latest`), but with a twist: we need to keep `test_clang_opt` because [we invoke it by name on FreeBSD](https://github.com/Mbed-TLS/mbedtls-test/blob/master/vars/common.groovy#L83).

(Alternatively maybe we could hide `test_clang_opt` on Linux? We can't just write a `support` function to skip it because the CI enforces that all components must run on at least one Linux image, to check that we don't accidentally skip a component. We could make the script only define the function `test_clang_opt` if not on Linux maybe? Or skip `test_clang_opt` from `--list-all`? Alternatively, we could create a `clang-latest` executable on FreeBSD in the Groovy script. But all of these solutions feel fiddly. Suggestions welcome.)

For the `latest` compilers, keep the different build options. For the `earliest` compilers, just do a `-O` build.

The new components will only run where the necessary compiler is installed, so each will need a corresponding `support_test_XXX` function that checks e.g.
```
support_test_clang_earliest () {
    type clang-earliest >/dev/null 2>/dev/null
}
```

For old compilers, I don't expect any trouble (they're only slightly older than what we test with now). For new compilers, I expect new warnings, so the bulk of the work in this task will likely to be to resolve those warnings.

Prerequisites:
* https://github.com/Mbed-TLS/mbedtls-test/issues/110
* https://github.com/Mbed-TLS/mbedtls/issues/7891

Likely follow-ups:

* Pass more warnings to clang, in particular `-Wdocumentation` (https://github.com/Mbed-TLS/mbedtls/issues/7886). We might want to change `test_build_opt` to use cmake.
* Switch other jobs to a newer compiler. We'll probably want to run most tests with the `latest` compiler versions, but we'll do that by setting environment variables in the CI scripts so that locally `all.sh` jobs use whatever compiler the developer wants.
* https://github.com/Mbed-TLS/mbedtls/issues/6610

### PR DESCRIPTION
**Pull request by gowthamsk-arm**
Closes #7891

Needs https://github.com/Mbed-TLS/mbedtls-test/pull/118 to be merged, otherwise the CI won't pass.

This PR has adds following changes: 

- The Ubuntu 16.04 and 22.04 docker images in mbedtls-test have been updated with earliest and latest versions of gcc and clang respectively. This PR adds the components and support functions required for the CI to run these compilers. For FreeBSD, we invoke the function by name so a condition is added to disable the existing test_clang_opt function for Linux. 

- Running clang-16 on mbedtls reports warnings of type "-Wstrict-prototypes". This patch fixes these warnings by adding void to functions with no arguments.

## PR checklist

- [x] **changelog**  not required
- [x] **backport** #7969
- [x] **tests** not required

### CODE REVIEW

**Comment by gowthamsk-arm** [comment]
This PR has a dependency on https://github.com/Mbed-TLS/mbedtls-test/pull/118
Internal CI run: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/509/

**Comment by gowthamsk-arm** [comment]
OpenCI run: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/15/

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:4014]: I would like the new components to print out the compiler version that they're using. Otherwise that information doesn't appear in the CI logs. I think a good way to have the information would be to run `$cc --version` in `test_build_opt`.

Maybe we should also add this information to `output_env.sh`.

File: tests/suites/test_suite_aes.function
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_aes.function:421]: I'd prefer to keep allowing `()` in test functions. It's a more logical syntax (which C lacks only due to backward compatibility constraint), and if we stop allowing it, we'll likely forget in new work and then get surprised when the CI fails with `clang-latest`. If we do, then we don't need the massive (in terms of changed lines) changes to `*.function`.

This only requires a small change in `tests/scripts/generate_test_code.py`: in `parse_function_code`, if `args` is blank (ideally also if `args` only contains comments, but that's not critical), add `void` to it.

File: tests/scripts/all.sh
Comment (gowthamsk-arm) [reply tests/scripts/all.sh:4014]: Will add them.

File: tests/suites/test_suite_aes.function
Comment (gowthamsk-arm) [reply tests/suites/test_suite_aes.function:421]: Makes sense. Will drop the `*.function` changes.

Comment (gowthamsk-arm) [reply tests/suites/test_suite_aes.function:421]: There is a condition in `tests/scripts/generate_test_code.py` in function `gen_from_test_data` 
`if len(test_args) != len(func_args):`
which checks if the test data and the function signature have the same number of arguments. 
This reports an error saying the function generated and the function parameter listed in `*.data` are not the same if I add `void` insertion in the script. :( 
Do you want me to update the script and all the test data files which use the void arg functions? :)

Comment (gowthamsk-arm) [reply tests/suites/test_suite_aes.function:421]: Either way, we will end up changing lots of files :)

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_aes.function:421]: There's already code in the script to accept `void` as meaning no arguments, since 47e2e8817d938bc02ec94be8f01a4755eacca4c5. Looking at it now (I'd forgotten how it was done), it seems that we're stripping `void` if it was present in the source, so we should unconditionally add it when generating. So yes, adding it in `gen_from_test_data` does sound good.

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_aes.function:421]: By the way, there are unit tests in `tests/scripts/test_generate_test_code.py`. If inserting `void` breaks a unit test, please update the expected result from that test. If inserting `void` doesn't break an existing unit test, please add one.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from one minor thing in `all.sh`.

File: tests/scripts/all.sh
Comment (gowthamsk-arm) [reply tests/scripts/all.sh:202]: Yes, will update this.

Comment (gowthamsk-arm) [reply tests/scripts/all.sh]: Will trigger a single CI run to verify this change.

Comment (gowthamsk-arm) [reply tests/scripts/all.sh]: Triggered: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/522/

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:4023]: Instead of running `clang-latest` and such, please run `"$CLANG_LATEST"` and such, and set them to default to their current names but allow changing them with command line options `--clang-latest` and such. The same thing we're doing with `--openssl-next` setting `OPENSSL_NEXT`, etc. And with the same environment overrides, while we're at it. (Which means that we don't actually need the symlinks on the CI, but we can handle that later.)

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Following our Slack discussion, I'm not going to ask for a redesign, but please add one convenience feature for local development (which has historical precedent).

**Comment by tgonzalezorlandoarm** [comment]
LGTM after resolving @gilles-peskine-arm comments

File: tests/scripts/all.sh
Comment (gowthamsk-arm) [reply tests/scripts/all.sh:4023]: Once this PR is merged, I will create a new PR on `mbedtls-test` removing the symlinks.

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh:4023]: More precisely: setting environment variables instead of creating symlinks.

Comment (gowthamsk-arm) [reply tests/scripts/all.sh:4023]: I'm planning to set `"$CLANG_LATEST"` to `clang-16` as the default value within `all.sh`. So do we still need to set environment variables ?

**Review by tgonzalezorlandoarm — APPROVED** [review]

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Just a minor thing

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:180]:
```suggestion
    : ${CLANG_EARLIEST:="clang-earliest"}
```
and likewise for gcc below.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM except for a typo that prevented the CI from running

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by tgonzalezorlandoarm — APPROVED** [review]

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:182]: Very minor: there used to be a blank line here, it's still present in the 2.28 branch but not here. Not worth changing if there's nothing else to change.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:4037]: Why `-O0`? That's the fastest build, but `-O` or -O2` is faster when doing build and test. Also, `-O0` omits some warnings.

Comment (gowthamsk-arm) [reply tests/scripts/all.sh:4037]: My bad, I looked into the changes thoroughly and this was supposed to be `-O`. :( 
Its also mentioned in #7891. 
> For the earliest compilers, just do a -O build.

Shall I create a new PR for this change?

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh:4037]: I'll fix it in https://github.com/Mbed-TLS/mbedtls/pull/9286 which is where I noticed it.

Comment (gowthamsk-arm) [reply tests/scripts/all.sh:4037]: Thank you.

**Pull request merged**
